keps/sig-storage/2644-honor-pv-reclaim-policy/README.md

```diff
@@ -0,0 +1,421 @@
+
+If when processing the PV update it is observed that it has `kubernetes.io/pv-deletion-protection` finalizer and
+`DeletionTimestamp` set, then the volume deletion request is forwarded to the driver(provided other pre-defined 
+conditions are met).
+
+
+
+### Risks and Mitigations
+
+The primary risk is volume deletion to a user that expect the current behavior, i.e, they do not expect the volume to be
+deleted from the underlying storage infrastructure when for a bound PV-PVC pair, the PV is deleted followed by deleting
+the PVC. This can be mitigated by initially introducing the fix behind a feature gate, followed by explicitly deprecating
+the current behavior.
+
+## Design Details
+
+When processing the PV update in `kube-controller-manager`, if it's observed that the PV is still bound to a valid claim,
+and also has `DeletionTimestamp` set, then a finalizer is added to the PV.
+
+```go
+// PvDeletionProtectionFinalizer finalizer is added to the PV when it is observed that the PV
+// has a deletion timestamp but still associated with a valid claim.
+PvDeletionProtectionFinalizer = "kubernetes.io/pv-deletion-protection"
+```
+
+```go
+// Function: syncVolume
+// Code above skipped for brevity...
+if !found {
+			klog.V(4).Infof("synchronizing PersistentVolume[%s]: claim %s not found", volume.Name, claimrefToClaimKey(volume.Spec.ClaimRef))
+			// Fall through with claim = nil
+		} else {
+			var ok bool
+			claim, ok = obj.(*v1.PersistentVolumeClaim)
+			if !ok {
+				return fmt.Errorf("Cannot convert object from volume cache to volume %q!?: %#v", claim.Spec.VolumeName, obj)
+			}
+			klog.V(4).Infof("synchronizing PersistentVolume[%s]: claim %s found: %s", volume.Name, claimrefToClaimKey(volume.Spec.ClaimRef), getClaimStatusForLogging(claim))
+			// Check if the PV has a deletion timestamp, if so, annotate the PV to indicate that the PV can be deleted
+			// from the underlying storage.
+			if protectionutil.NeedToAddFinalizer(volume, pvutil.PvDeletionProtectionFinalizer) {
+				klog.Infof("synchronizing PersistentVolume[%s]: the volume was requested to be deleted but still" +
+					" it is still associated with the claim %s", volume.Name, claimrefToClaimKey(volume.Spec.ClaimRef))
+				_, err = ctrl.addPvDeletionProtectionFinalizer(volume)
+				if err != nil {
+					klog.V(4).Infof("synchronizing PersistentVolume[%s]: failed to add the pv deletion protection finalizer", volume.Name)
+				}
+				klog.Infof("successfully added pv deletion protection finalizer to indicate the volume is to be deleted.")
+			}
+		}
+// Remaining code below skipped for brevity...
+// Please refer kubernetes/kubernetes/pkg/controller/volume/persistentvolume/pv_controller.go for the full code.
+```
+
+When `deleteVolumeOperation` executes, the finalizer is checked, if present it implies that the PV was deleted first,
+in this scenario `DeletionTimestamp` checks can be ignored. Once the volume is deleted by the plugin the finalizer can
+be safely removed.
```

Comment:
As suggested above, the PV deletion finalizer should be always present. It does not tell if PV was deleted before PVC. On the bright side, nobody should worry about what was deleted first. In both cases, "Once the volume is deleted by the plugin the finalizer can be safely removed."